lib/eslint.js

```diff
@@ -1036,8 +1036,8 @@ module.exports = (function() {
 
             // if current node is function declaration, add it to the list
             var current = controller.current();
-            if (["FunctionDeclaration", "FunctionExpression",
-                    "ArrowFunctionExpression", "SwitchStatement"].indexOf(current.type) >= 0) {
+            if (["FunctionDeclaration", "FunctionExpression", "ArrowFunctionExpression"].indexOf(current.type) >= 0 ||
+                    currentConfig.ecmaFeatures.blockBindings) {
```

Comment:
I'm not sure this is correct. Wouldn't you only want to do this if `blockBindings` is true and also the node type is `BlockStatement`?